### ISSUE TITLE
[Optim] Optimize GroupNormSilu Performance via Precision Casting

### ISSUE THREAD

**Issue by zhangtao0408** [zhangtao0408/SeedVR#9]
# Problem Description

<img width="1814" height="434" alt="Image" src="https://github.com/user-attachments/assets/bd353b8c-b0c4-4020-b8b1-33d044107a5c" />

The GroupNormSiluoperator currently runs in Float32 (FP32) precision during inference, accounting for 4.6%​ of total network runtime. This represents a significant optimization opportunity, as lower precision (e.g., Float16/BFloat16) could provide substantial performance gains on supported hardware without meaningful accuracy loss.

# Current Status
<img width="2888" height="194" alt="Image" src="https://github.com/user-attachments/assets/9561ed3a-239b-4e7d-836b-ed4eca92b600" />

<img width="978" height="142" alt="Image" src="https://github.com/user-attachments/assets/3ebcdeab-07b9-4dc5-af35-3f2f84890434" />

- Runtime Share: GroupNormSilu operations consume 4.6% of total inference time
- Precision: Currently running exclusively in Float32 (FP32) precision
- Hardware Support: Target hardware (NPU) supports efficient lower-precision computation
- Impact: Unnecessarily high precision for activation functions limits inference speed

# Optimization Goals

- Reduce GroupNormSilu execution time by leveraging lower precision computation
- Maintain model accuracy
- Minimize precision conversion overhead
- Achieve measurable inference speedup targeting 2-3% total network improvement

# Possible Solution

>https://www.hiascend.com/document/detail/zh/Pytorch/720/apiref/torchnpuCustomsapi/context/torch_npu-npu_group_norm_silu.md

Cast GroupNormSilu inputs/outputs to lower precision (Float16/BFloat16) by npu fusion operator `npu_group_norm_silu.md`

```python
# Current implementation
for i, (w, b) in enumerate(zip(weights, biases)):
      x[i] = F.group_norm(x[i], num_groups_per_chunk, w, b, norm_layer.eps)
      x[i] = x[i].to(input_dtype)
x = torch.cat(x, dim=1)

# Optimized version
for i, (w, b) in enumerate(zip(weights, biases)):
      if npu_available:
          _x_input = x[i].to(torch.bfloat16)
          _w_input = w.to(torch.bfloat16)
          _b_input = b.to(torch.bfloat16)
          x[i] = torch_npu.npu_group_norm_silu(_x_input, _w_input, _b_input, group=num_groups_per_chunk, eps=norm_layer.eps)
      else:
          x[i] = F.group_norm(x[i], num_groups_per_chunk, w, b, norm_layer.eps)
      x[i] = x[i].to(input_dtype)
x = torch.cat(x, dim=1)

```

**Comment by zhangtao0408** [comment]
1. No group norm
```shell
Total time: 30.030997276306152 seconds, 
Preprocess time: 1.2874603271484375e-05 seconds,
Encode time: 0.5221023559570312 seconds, 
DIT time: 27.184649229049683 seconds, 
Post process time: 2.3242266178131104 seconds
```

2. Add Group Norm
```shell
Total time: 29.474334001541138 seconds, 
Preprocess time: 2.2172927856445312e-05 seconds, 
Encode time: 0.5889928340911865 seconds, 
DIT time: 26.617782592773438 seconds, 
Post process time: 2.2675299644470215 seconds
```

DiT -> 2%, but has precision error.


3. Videos
[no_group_norm_silu.mp4](https://github.com/user-attachments/assets/c539df6d-ce65-4b23-b54f-82390ea1926d)
[with_group_norm_silu.mp4](https://github.com/user-attachments/assets/d07d74fc-5731-4e7b-b267-47a6b241ade6)

**Comment by zhangtao0408** [comment]
The initial analysis had a misunderstanding. In the DIT module, there are two types of group normalization:

1. Group Norm without SILU
> `models/video_vae_v3/modules/attn_video_vae.py`
```python
class Upsample3D(Upsample2D):
    ...

    def forward(
        self,
        hidden_states: torch.FloatTensor,
        output_size: Optional[int] = None,
        memory_state: MemoryState = MemoryState.DISABLED,
        **kwargs,
    ) -> torch.FloatTensor:
        ...

        if hasattr(self, "norm") and self.norm is not None:
            # [Overridden] change to causal norm.
            hidden_states = causal_norm_wrapper(self.norm, hidden_states)

        ...
```

2. Group Norm with SILU
> `models/video_vae_v3/modules/video_vae.py`
```python
class Encoder3D(nn.Module):
    ...
    
    def forward(self, sample: torch.FloatTensor, memory_state: MemoryState) -> torch.FloatTensor:
        ...

        # post-process
        sample = causal_norm_wrapper(self.conv_norm_out, sample)    # group_norm
        sample = self.conv_act(sample)                              # silu
        sample = self.conv_out(sample, memory_state=memory_state)

        return sample
```

<img width="833" height="369" alt="Image" src="https://github.com/user-attachments/assets/68d2591b-fa9a-43b7-8741-91f02cf3890c" />

According to the [official documentation](https://www.hiascend.com/document/detail/zh/Pytorch/720/apiref/torchnpuCustomsapi/context/torch_npu-npu_group_norm_silu.md), only the second case (Group Norm followed by SILU) is eligible for the `torch_npu.npu_group_norm_silu` fusion operator. That's why there was a precision bug while using the `npu_group_norm_silu` op.

We will start another issue for the 2nd case. Let's go on for the 1st case.

**Comment by zhangtao0408** [comment]
# Condition #1: Group Norm without SILU

<img width="2654" height="290" alt="Image" src="https://github.com/user-attachments/assets/4b60176f-58f5-49aa-80ea-a8d53fb00b25" />

As shown in the image above, there are three `cast` operators above and under `group_norm`. According to the code in [torch_npu/csrc/aten/AutoCastOps.cpp](https://gitcode.com/Ascend/pytorch/blob/v7.2.0-pytorch2.6.0/torch_npu/csrc/aten/AutoCastOps.cpp):

```c++
KERNEL_PRIVATEUSEONE(group_norm, fp32)
```

Whenever AutoCast is invoked, the input parameters of `group_norm` are forced to be converted to FP32 through autocast.

However, this creates an issue: our initialization code uses autocast to globally convert the inference precision to bfloat16. This causes `group_norm` to first convert from FP32 to BF16, and then back to FP32 for inference, resulting in an extra conversion step. The relevant code is as follows:

```python
with torch.no_grad(), torch.autocast("cuda", torch.bfloat16, enabled=True):
    video_tensors = runner.inference(
        noises=noises,
        conditions=conditions,
        dit_offload=dit_offload,
        **text_embeds_dict,
    )
```

Therefore, we need to eliminate the overhead introduced by autocast. A sample approach to reduce the redundant cast operations is shown below:

```python
# Current implementation
for i, (w, b) in enumerate(zip(weights, biases)):
    # Disable Autocast
    with torch.amp.autocast('npu', enabled=False):
        x[i] = F.group_norm(x[i], num_groups_per_chunk, w, b, norm_layer.eps)
    x[i] = x[i].to(input_dtype)
x = torch.cat(x, dim=1)
```

# Demo Results

<img width="2066" height="242" alt="Image" src="https://github.com/user-attachments/assets/8d1a26b9-814f-480a-b739-a25822826bcc" />

By a demo, we disabled the autocast while processing `group_norm`, we can see the number of cast reduced 42%, `cast` and `group_norm` ratio reduced to 2~3% at e2e.

<img width="3449" height="242" alt="Image" src="https://github.com/user-attachments/assets/201143ef-9b8b-4ad1-bf23-c474c36826af" />

Additionally, comparing the performance of the group norm operator between bfloat16 and float32 precision, the duration time for bfloat16 is approximately 1.7–1.8x faster than that of float32.

Group Norm Profiling: https://pan.quark.cn/s/9ce6da469a77
GrpupNormOpsInfo: https://www.kdocs.cn/l/caVGqTdHsq1L
